Remotion/ObjectBinding/Web.ClientScript/Scripts/BocReferenceValueBase.ts

```diff
@@ -0,0 +1,223 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the 
+// GNU Lesser General Public License for more details.
+// 
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+// 
+
+type ResourceIdentifier = {
+  LoadIconFailedErrorMessage: string,
+  LoadDataFailedErrorMessage: string,
+  SearchStringForDropDownDoesNotMatchRegexMessage: string,
+  NoDataFoundMessage: string,
+  InternetExplorerScreenReaderLabelText: string
+}
+
+type SearchStringValidationInfo = {
+  ValidSearchStringRegex: string,
+  ValidSearchStringForDropDownRegex: string,
+  SearchStringForDropDownDoesNotMatchRegexMessage: string,
+  IgnoreSearchStringForDropDownUponValidInput: string
+}
+
+type CommandInfo = {
+  href: Nullable<string>,
+  target: Nullable<string>,
+  onClick: Nullable<string>,
+  title: Nullable<string>
+}
+
+type IconContext = {
+  businessObjectClass: Nullable<string>,
+  arguments: Nullable<string>
+}
+
+type ErrorHandler = (error: Sys.Net.WebServiceError) => void
+
+type IconInformation = {
+  Url: string,
+  AlternateText: string,
+  ToolTip: string,
+  Width: number,
+  Height: number
+}
+
+class BocReferenceValueBase
+{
+  private static _nullIconUrl: Nullable<string>
+
+  public static InitializeGlobals (nullIconUrl: string): void
+  {
+    BocReferenceValueBase._nullIconUrl = nullIconUrl;
+  };
+
+  public static UpdateCommand (oldCommand: JQuery, businessObject: any, isIconUpdateEnabled: boolean, controlServiceUrl: Nullable<string>, iconContext: Nullable<IconContext>, commandInfo: Nullable<CommandInfo>, onFailure: ErrorHandler): JQuery
+  {
+    ArgumentUtility.CheckNotNull('oldCommand', oldCommand);
+    ArgumentUtility.CheckTypeIsString('businessObject', businessObject);
+    ArgumentUtility.CheckNotNullAndTypeIsBoolean('isIconUpdateEnabled', isIconUpdateEnabled);
+    if (isIconUpdateEnabled)
+      ArgumentUtility.CheckNotNullAndTypeIsString('controlServiceUrl', controlServiceUrl);
+    if (isIconUpdateEnabled)
+      ArgumentUtility.CheckNotNullAndTypeIsObject('iconContext', iconContext);
+    ArgumentUtility.CheckTypeIsObject('commandInfo', commandInfo);
+    ArgumentUtility.CheckTypeIsFunction('onFailure', onFailure);
+
+    var newCommand = BocReferenceValueBase.CreateCommand(oldCommand, commandInfo, businessObject);
+
+    var oldIcon = oldCommand.find('img');
+    var newIcon = BocReferenceValueBase.CreateEmptyIcon(oldIcon, newCommand.prop('title'));
+    newCommand.append(newIcon);
+
+    oldCommand.replaceWith(newCommand);
+
+    if (isIconUpdateEnabled)
+    {
+      var pageRequestManager = Sys.WebForms.PageRequestManager.getInstance();
+      if (pageRequestManager.get_isInAsyncPostBack())
+        return oldCommand;
+
+      var params: any = { businessObject: businessObject };
+      for (var propertyName in iconContext)
+        params[propertyName] = (iconContext as any)[propertyName];
+
+      Sys.Net.WebServiceProxy.invoke(
```

Comment:
Use WebServiceProxyInvokeFunction<TResult, TUserContext> as type. See Utilities.WebServiceUtility.

Also, TODO for changing implementation to use WebServiceUtility.Execute instead of doing the calls manually here and in BocAutoCompleteReferenceValue.jquery.js. I'd almost say that we change it immediately but it's just a bit out of scope and we don't really have that much extra work just fixing the typings right now.